advancedtls: add new module for advanced TLS handshaker

ZhenLian opened the pull request:
This PR contains the implementation and tests of a utility library for some advanced TLS features, including:

1. Credential Reloading: users are able to reload their credentials without restarting the app. Note that the "credential" here can be a private key, its corresponding certificate bundle, or trusted CA certificate bundle.

2. Server Authorization: on client side, we let users specify their own functions to check peers' identities when handshakes are finished. This feature is sometimes also referred to as "secure name checking" or "hostname verification".

There are some previous discussions in this deprecated PR: https://github.com/grpc/grpc-go/pull/3044. The goal of this PR remains the same, but change the APIs to provide better functionalities. 
This PR includes the implementation and unit tests and integration tests for the client and server without credential reloading. Next PR will cover the integration tests for credential reloading.

ZhenLian commented:
@jiangtaoli2016 Could you please take a first round of review on the API design? Thank you so much for the help!

jiangtaoli2016 commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,375 @@
+// will be used for read/write.
+//
+// This function returns newConn if rawConn doesn't implement syscall.Conn.
+func WrapSyscallConn(rawConn, newConn net.Conn) net.Conn {
```
Can grpc-go expose this function to public, so that we don't have to copy the implementation?
@dfawley

ZhenLian replied:
Yes, and another alternative might be we keep this function for now, and depend on the exposed one later, if it doesn't raise any other concerns.

jiangtaoli2016 commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,375 @@
+		RootCAs:              o.RootCACerts,
+	}
+	// We fall back to normal config settings if users don't need any of the advanced TLS features.
+	if o.GetClientCertificate == nil && o.GetRootCAs == nil && o.VerifyPeer == nil {
```
Why not always set InsecureSkipVerify = true?
In any case, VerifyPeerCertificate from buildVerifyFunc will be called.

ZhenLian replied:
Done.

jiangtaoli2016 commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,375 @@
+		}, nil
+	}
+	if o.GetRootCAs == nil && o.RootCACerts == nil {
+		return nil, fmt.Errorf("server needs trust certs if using mutual TLS")
```
nit: may reword as "server needs to provide Root CA certs if using mutual TLS"?

jiangtaoli2016 commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,375 @@
+	// use local cfg to avoid clobbering ServerName if using multiple endpoints
+	cfg := cloneTLSConfig(c.config)
+	if cfg.ServerName == "" {
+		colonPos := strings.LastIndex(authority, ":")
```
why you strip character after ":"? It is better to pass authority as it is to CustomVerificationFunc, so that application can decide? It is not gRPC transport credentials library's job to strip the trailing port.

ZhenLian replied:
Yeah I agree. We could pass the whole authority to user for custom stripping.

jiangtaoli2016 commented on security/advancedtls/advancedtls_test.go:
```diff
@@ -0,0 +1,482 @@
+		serverRoot                 *x509.CertPool
+		serverGetRoot              func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error)
+		serverExpectError          bool
+	}{
```
We may want to test the cases where the root CA certs do not match the peer certificates. E.g., server provides a valid certificate, but server's CA cert is different from client Root cert.

ZhenLian replied:
Added the following test cases:

- Client_reload_both_certs_verifyFunc_Server_reload_both_certs_mutualTLS
- Client_wrong_peer_cert_Server_reload_both_certs_mutualTLS
- Client_wrong_peer_cert_Server_reload_both_certs_mutualTLS
- Client_reload_both_certs_verifyFunc_Server_wrong_peer_cert
- Client_reload_both_certs_verifyFunc_Server_wrong_trust_cert

jiangtaoli2016 commented:
In current implementation, client has to provide either RootCA or GetRootCA. 

We see use case https://github.com/grpc/grpc/pull/20530 that client does not know root CA certs in advance. In that case, we give the power to client to validate via VerifyPeer function (since application can access target name and server certain chain).

Thus, if client does not provide RootCA or GetRootCA, instead of failing, we just don't validate the server cert and let application decide via VerifyPeer. @ZhenLian Could update your implementation as well?

jiangtaoli2016 commented:
Thanks @ZhenLian for the implementation. 
My knowledge in Golang is limited. 
@cesarghali Please check Go readability and security suggestions.
@dfawley Please review on API and make sure you are comfortable.

ZhenLian commented:
The API now is updated as:
When the client side specifies neither RootCA nor GetRootCA but with VerifyPeer, we will only use VerifyPeer to see if this RPC call is valid.
Note that is client specifies none of the three, we will still fall back to default checks, which are certificate check and host name check.

cesarghali commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,375 @@
+	credentials.AuthInfo, error) {
+	cfg := cloneTLSConfig(c.config)
+	// We build server side verification function only when root cert reloading is needed.
+	if c.getRootCAs != nil {
```
At first I thought that using `MutualAuth` here would be a more clear indication that we need to set the verifier since `MutualAuth` means that the server will get a client cert that it needs to verify. But looking at `ServerOptions.config`, it seems that `MutualAuth` doesn't always mean verify, it might mean `RequireAnyClientCert` only. Can you please add a comment to clarify what `c.getRootCAs != nil` means here?

ZhenLian replied:
Yes, the behaviors are:

if server want client cert(mTLS): 
        if c.getRootCAs != nil:
                clientAuth = RequireAnyClientCert, because we have to disable the default check, 
                as mentioned in last comment
        else:
                clientAuth = RequireAndVerify, because basically user didn't specify anything 
                special for server side, and thus we fall back to normal check(also includes 
                hostname check)
        else:
                clientAuth = NoCert

cesarghali replied:
I think it's still worth adding a comment about this in the code. It doesn't have to be here, it can be in the config function, basically combined with the comment I think we should add in that function.

cesarghali commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,375 @@
+func buildVerifyFunc(c *advancedTLSCreds,
+	serverName string,
+	rawConn net.Conn) func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	verifyFunc := func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
```
You can simply do `return func(rawCerts...` instead of defining a variable.

cesarghali commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,375 @@
+		// instead of failing, we just don't validate the server cert and let
+		// application decide via VerifyPeer
+		if c.isClient && c.config.RootCAs == nil && c.getRootCAs == nil {
+			if c.isClient && c.verifyFunc != nil {
```
There's no need to check `c.isClient` since the outer `if` takes care of that.

cesarghali commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,375 @@
+		// verify peers' certificates against RootCAs and get verifiedChains
+		certs := make([]*x509.Certificate, len(rawCerts))
+		for i, asn1Data := range rawCerts {
+			cert, _ := x509.ParseCertificate(asn1Data)
```
Always handle returned errors.

cesarghali commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,375 @@
+			Intermediates: x509.NewCertPool(),
+		}
+		if !c.isClient {
+			opts.KeyUsages = []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth}
```
Do we need to use `ExtKeyUsageServerAuth` for clients?

ZhenLian replied:
I am not sure... theoretically we need to, but in the implementation here(https://github.com/golang/go/blob/master/src/crypto/tls/handshake_client.go#L804-L820), ExtKeyUsageServerAuth is not set, while for server side, the equivalent field is set(https://github.com/golang/go/blob/master/src/crypto/tls/handshake_server.go#L753).

I added that for now and let's see :-)

cesarghali replied:
SG.

cesarghali commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,375 @@
+}
+
+// TODO(ZhenLian): The code below are duplicates with gRPC-Go under
+// TODO(ZhenLian): credentials/internal. Consider refactoring in the future.
```
You can remove `// TODO(ZhenLian):` from this line. Feel free to add spaces instead if you want to align the TODO text.

ZhenLian commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,375 @@
+	if o.GetRootCAs == nil && o.RootCACerts == nil {
+		return nil, fmt.Errorf("server needs to provide root CA certs if using mutual TLS")
+	}
+	// We fall back to normal config settings if users don't need to reload root certificates.
```
We does root credential reloading by rewriting the verification function, and disabling the default check.  We have to disable the default check because otherwise, every time it will use the same static certificate, making reloading impossible. 
On server side, to disable the default check, we have to deliberately set ClientAuth to "NoClientCert"(btw, the equivalent part for client side is to set InSecureSkipVerify to true), which will not do any default check, and then perform custom reloading in the following VerifyPeerCertificate function. 
We have contacted the Go team and this is their suggested way to do root credential reloading. It doesn't look clean enough, and they don't have any plans to change their APIs. But on the other hand, this also proves the utility libraries are beneficial to some users since this might be tricky :-)

cesarghali replied:
If it's still relevant, it might be worth adding a comment to summarize this :)

ZhenLian commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,375 @@
+}
+
+// ClientOptions contains all the fields and functions needed to be filled by the client.
+type ClientOptions struct {
```
I've thought of some ways of extraction, but the input parameters of GetClientCertificate(for client) and GetCertificate(for server) are different. It might look weird if we extract the other 3 and call them "CertificateOptions" ... Do we have any better ways for this?

cesarghali replied:
I was thinking of something like the following:

```
type CertificateOptions struct {
    // Options common for clients and servers.
}

type ClientOptions struct {
    CertOptions CertificateOptions
    // The rest of client options.
}

type ServerOptions struct {
    CertOptions CertificateOptions
    // The rest of server options.
}
```

ZhenLian commented:
@cesarghali Thank you so much for the review! I left a few more comments. Please let me know if anything could be further improved :-)

ZhenLian commented:
@cesarghali 
Hi Cesar, I've updated the comments inside config() functions and refactored the Options struct as we discussed. Could you please take a look at it(as well as the tests) to see if there are any other things that could be improved? Thank you so much!

ZhenLian commented:
Friendly ping :-)

cesarghali commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,391 @@
+	// If GetRootCAs is set and RootCACerts is nil, the server will invoke this function every time
+	// asked to check certificates sent from the client when a new connection is established.
+	// This is known as root CA certificate reloading.
+	GetRootCAs func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error)
```
Can we also use `RootCertificateOptions` instead of `RootCACerts` and `GetRootCAs`?

ZhenLian commented:
@dfawley
Hi Doug, sorry for the delay of this PR. Now our team has completed the review process. Could you please take a look at the overall structure, the go module files, etc? Thanks in advance for your time!

ZhenLian commented:
@dfawley friendly ping :-)

dfawley commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,377 @@
+// Package advancedtls is a utility library containing functions to construct
+// credentials.TransportCredentials that can perform credential reloading and custom
+// server authorization.
+// All APIs in this package are EXPERIMENTAL.
```
This is not necessary since the module is v0 by default.

dfawley commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,377 @@
+	"google.golang.org/grpc/credentials"
+)
+
+type sysConn = syscall.Conn
```
Worth a comment here why this is needed.  Also, I'd move this down to where it's needed.

dfawley commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,377 @@
+// CustomVerificationFunc is the function defined by users to perform custom server
+// authorization.
+type CustomVerificationFunc func(serverName string,
+	rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error
```
Unless you are sure you will never need new parameters here, it's better to pass these as fields in a struct.  Also, if this function might every potentially want to return information, it should also return a struct type (which would be empty for now).

dfawley commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,377 @@
+	// If GetRootCAs is set and RootCACerts is nil, the client will invoke this function every time
+	// asked to check certificates sent from the server when a new connection is established.
+	// This is known as root CA certificate reloading.
+	GetRootCAs func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error)
```
Similar comments as above regarding extensibility.

dfawley commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,377 @@
+	// RootCACerts or GetRootCAs indicates the the certificates trusted by the client side. The rules
+	// for setting these two fields are:
+	// Either RootCACerts or GetRootCAs must be set; the other will be ignored
+	//----------------------------------------------------------------------------------------------
```
This comment looks like it belongs with the struct and not the RootCACerts field.

dfawley commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,377 @@
+// It is used by both ClientOptions and ServerOptions.
+type RootCertificateOptions struct {
+	// ---------------General Rules For Root Certificate Setting------------------------------------
+	// RootCACerts or GetRootCAs indicates the the certificates trusted by the client side. The rules
```
"client" --- do you mean "local" in the event this is in ServerOptions?

dfawley commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,377 @@
+	//     Either Certificates or GetClientCertificate must be set; the other will be ignored
+	// Otherwise:
+	//     Nothing needed(the two fields will be ignored)
+	//----------------------------------------------------------------------------------------------
```
Similar: I think if a comment generally covers multiple fields then it should be with the struct and not any individual field.

dfawley commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,377 @@
+}
+
+func (c *advancedTLSCreds) ClientHandshake(ctx context.Context, authority string,
+	rawConn net.Conn) (_ net.Conn, _ credentials.AuthInfo, err error) {
```
Nit: I wouldn't name return values at all unless you need to or they help with the documentation.

Also, I'd unwrap the line as long lines are fine in Go and this isn't that excessive.

dfawley commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,377 @@
+}
+
+func (c *advancedTLSCreds) ServerHandshake(rawConn net.Conn) (net.Conn,
+	credentials.AuthInfo, error) {
```
Unwrap this line?

dfawley commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,377 @@
+			return nil, nil, err
+		}
+	case <-ctx.Done():
+		return nil, nil, ctx.Err()
```
`conn.Close()`

dfawley commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,377 @@
+	select {
+	case err := <-errChannel:
+		if err != nil {
+			return nil, nil, err
```
`conn.Close()`?

dfawley commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,377 @@
+	errChannel := make(chan error, 1)
+	go func() {
+		errChannel <- conn.Handshake()
+	}()
```
`close(errChannel)` is unnecessary but probably a good idea.

dfawley commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,377 @@
+	}
+	conn := tls.Server(rawConn, cfg)
+	if err := conn.Handshake(); err != nil {
+		return nil, nil, err
```
`conn.Close()`?

dfawley commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,377 @@
+		for _, cert := range certs[1:] {
+			opts.Intermediates.AddCert(cert)
+		}
+		// We use default hostname check if users doesn't specify verifyFunc function
```
Nit: users->user (or doesn't->don't)

dfawley commented on security/advancedtls/advancedtls.go:
```diff
@@ -0,0 +1,377 @@
+	}
+}
+
+// NewClientTLS uses ClientOptions to construct a TransportCredentials based on TLS.
```
This is a "tls" package - is "TLS" necessary in the name here?

ZhenLian commented:
@dfawley 
Hi Doug, thank you so much for the suggestions and comments. I have addressed them all. Could you please take a second look?
By the way, I made another PR, https://github.com/grpc/grpc-go/pull/3300, to fix a small issue found here but might also apply to that place. It would be good if you could take a look as well.
Thanks again for the help :-)
